File: 49-group-anagrams/group-anagrams.java
import java.util.*;

class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {
        
        // Check for empty inputs
        if (strs == null || strs.length == 0)
            return new ArrayList<>();
        
        Map<String, List<String>> frequencyStringsMap = new HashMap<>();
        
        for (String str : strs) {
            
            String frequencyString = getFrequencyString(str);
            
            // If the frequency string is present, add the string to the list
            if (frequencyStringsMap.containsKey(frequencyString)) {
                frequencyStringsMap.get(frequencyString).add(str);
            } else {
                // else create a new list
                List<String> strList = new ArrayList<>();
                strList.add(str);
                frequencyStringsMap.put(frequencyString, strList);
            }
        }
        
        return new ArrayList<>(frequencyStringsMap.values());
    }
    
    private String getFrequencyString(String str) {
        int[] count = new int[26];
        
        for (char c : str.toCharArray()) {
            count[c - 'a']++;
        }
        
        return Arrays.toString(count);
    }
}

File: 49-group-anagrams/group-anagrams.cpp
class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        vector<vector<string>> ans;
        unordered_map<string,vector<string>> mp;
        for(auto &x: strs){
            vector<int> count(26, 0);
            for(auto c: x){
                count[c-'a']++;
            }
            string key = "";
            for(int i=0;i<26;i++){
                key+= "#" + to_string(count[i]);
            }
            mp[key].push_back(x);
        }
        for(auto& pair:mp){
            ans.push_back(move(pair.second));
        }
        return ans;
    }
};